chapter/1/rpc-reviewed-dont-keep.md

```diff
@@ -0,0 +1,619 @@
+parties. GridSolve, on the other hand, is relatively complex and provides a
+middleware for the communications using a client-agent-server model.
+
+#### Mobile Systems and Computation Offloading
+
+Mobile systems have become very powerful these days. With multi-core processors
+and gigabytes of RAM, they can undertake relatively complex computations without
+a hassle. Due to this advancement, they consume a larger amount of energy and
+hence, their batteries, despite becoming larger, drain quickly with usage.
+Moreover, mobile data (network bandwidth) is still limited and expensive. Due to
+these requirements, it's better to offload mobile computations from mobile
+systems when possible. RPC plays an important role in the communication for this
+*computation offloading*. Some of these services use Grid RPC technologies to
+offload this computation. Whereas, other technologies use an RMI(Remote Method
+Invocation) system for this.
+
+The Ibis Project {% cite ibis --file rpc %} builds an RMI and GMI (Group Method
+Invocation) model to facilitate outsourcing computation. Cuckoo {% cite cuckoo
+--file rpc %} uses this Ibis communication middleware to offload computation.
+
+The Microsoft's MAUI Project {% cite maui --file rpc %} uses RPC communication
+and allows partitioning of .NET applications and "fine-grained code offload to
+maximize energy savings with minimal burden on the programmer". MAUI decides the
+methods to offload to the external MAUI server at runtime.
+
+#### Async RPC, Futures and Promises
+
+Remote Procedure Calls can be asynchronous. Not only that but these async RPCs
+play in integral role in the *futures* and *promises*. *Future* and *promises*
+are programming constructs that where a *future* is seen as variable/data/return
+type/error while a *promise* is seen as a *future* that doesn't have a value,
+yet. We follow Finagle's {% cite finagle --file rpc %} definition of *futures*
+and *promises*, where the *promise* of a *future*(an empty *future*) is
+considered as a *request* while the async fulfillment of this *promise* by a
+*future* is seen as the *response*. This construct is primarily used for
+concurrent programming.
+
+Perhaps the most renowned systems using this type of RPC model are Twitter's
+Finagle{% cite finagle --file rpc %} and Cap'n Proto{% cite capnproto --file rpc %}.
+
+#### RPC in Microservices Ecosystem:
+
+RPC implementations have moved from a one-server model to multiple servers and
+on to dynamically-created, load-balanced microservices. RPC started as a
+separate implementations of REST, Streaming RPC, MAUI, gRPC, Cap'n Proto, and
+has now made it possible for integration of all these implementations as a
+single abstraction as a user *endpoint* service. The endpoints are the building
+blocks of *microservices*. These *microservices* interact with each other and
+applications and combine to give the feel of one large monolithic service.
```

Comment:
Maybe you should define what a single microservice is? Usually one "service" with a very simple, well-defined purpose, written in almost any language, that interacts with other microservices to give the feeling of a single monolithic service.